File: python/Sorcerer's Spell.py
def prime(k,n):
	while n >= 2:
		if k % n == 0:
			return n
		else:
			n-=1
	return 1

def gcd(a,b):
	if a ==0:
		return b;
	return gcd(b % a, a);


def findGCD(arr,n):
	result = arr[0];
	for i in range(1,n):
		result = gcd(arr[i],result)
		if(result == 1):
			return 1
	return result

if __name__ == "__main__":
	t = int(input())
	while(t):
		n, m = [int(a) for a in input().split()]
		odd = 0
		even = 0
		p = [int(a) for a in input().split()]
		o = findGCD(p, m)
		if o<=n:
			print(n-o)
		else:
			p = prime(o,n)
			print(n-p)
		t-=1

File: cpp/Sorcerer's Spell.cpp
#include <iostream>

using namespace std;

long long prime(long long k, long long n)
{
    long long o = n;
    while (o >= 2)
    {
        if (k % o == 0)
        {
            return o;
        }
        else
        {
            o--;
        }
    }
    return 1;
}

long long gcd(long long a, long long b)
{
    if (a == 0)
        return b;
    return gcd(b % a, a);
}

long long findGCD(long long arr[], long long n)
{
    long long result = arr[0];
    for (long long i = 1; i < n; i++)
    {
        result = gcd(arr[i], result);
        if (result == 1)
        {
            return 1;
        }
    }
    return result;
}

int main()
{
    long long t;
    cin >> t;
    while (t--)
    {
        long long n, m;
        bool odd = 0, even = 0;
        cin >> n >> m;
        long long p[m];
        for (long long i = 0; i < m; i++)
        {
            cin >> p[i];
        }
        long long o = findGCD(p, m);
        if (o <= n)
        {
            cout << n - o;
        }
        else
        {
            long long p = prime(o, n);
            cout << n - p;
        }
        cout << "\n";
    }
    return 0;
}